### PR TITLE
apply localityLbSetting to sni-dnat clusters

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

When debuging Replicated control planes multi-clusters, we found the locality loadbalancer not take effect. Actually we did not set it for sni-dnat clusters at all.

cc @venilnoronha  

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -93,17 +93,11 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
 
 	outboundClusters := configgen.buildOutboundClusters(env, proxy, push)
 
-	if env.Mesh.LocalityLbSetting != nil {
-		// apply load balancer setting fot cluster endpoints
-		applyLocalityLBSetting(proxy.Locality, outboundClusters, env.Mesh.LocalityLbSetting)
-	}
-	// Add a blackhole and passthrough cluster for catching traffic to unresolved routes
-	// DO NOT CALL PLUGINS for these two clusters.
-	outboundClusters = append(outboundClusters, buildBlackHoleCluster(env), buildDefaultPassthroughCluster(env, proxy))
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: I moved these 2 clusters to the last, we should not mutate them.

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -93,17 +93,11 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
...
-
 	switch proxy.Type {
 	case model.SidecarProxy:
 		outboundClusters = envoyfilter.ApplyClusterPatches(networking.EnvoyFilter_SIDECAR_OUTBOUND, proxy, push, outboundClusters)
+		// apply load balancer setting fot cluster endpoints
+		applyLocalityLBSetting(proxy.Locality, outboundClusters, env.Mesh.LocalityLbSetting)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: The `env.Mesh.LocalityLbSetting` is in the `applyLocalityLBSetting`.

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Should we do that patch before or after applying locality? Seems like after?

File: pilot/pkg/networking/core/v1alpha3/cluster_test.go [hunk 3]
```diff
@@ -859,6 +859,69 @@ func TestLocalityLB(t *testing.T) {
...
+				if locality.Region == "region1" && locality.SubZone == "subzone1" {
+					g.Expect(localityLbEndpoint.LoadBalancingWeight.GetValue()).To(Equal(uint32(34)))
+					g.Expect(localityLbEndpoint.LbEndpoints[0].LoadBalancingWeight.GetValue()).To(Equal(uint32(40)))
+				} else if locality.Region == "region1" && locality.SubZone == "subzone2" {
+					g.Expect(localityLbEndpoint.LoadBalancingWeight.GetValue()).To(Equal(uint32(17)))
+					g.Expect(localityLbEndpoint.LbEndpoints[0].LoadBalancingWeight.GetValue()).To(Equal(uint32(20)))
+				} else if locality.Region == "region2" {
+					g.Expect(localityLbEndpoint.LoadBalancingWeight.GetValue()).To(Equal(uint32(50)))
+					g.Expect(len(localityLbEndpoint.LbEndpoints)).To(Equal(1))
+					g.Expect(localityLbEndpoint.LbEndpoints[0].LoadBalancingWeight.GetValue()).To(Equal(uint32(40)))
```
Comment (venilnoronha) [inline pilot/pkg/networking/core/v1alpha3/cluster_test.go]: Should we also test the `priority` for each endpoint here?

File: pilot/pkg/networking/core/v1alpha3/cluster_test.go [hunk 4]
```diff
@@ -859,6 +859,69 @@ func TestLocalityLB(t *testing.T) {
...
+				}
+			}
+			return
+		}
+	}
+	t.Errorf("Do not generate sni-dnat cluster")
```
Comment (venilnoronha) [inline pilot/pkg/networking/core/v1alpha3/cluster_test.go]: s/Do/Did

**Comment by venilnoronha** [comment]
Issue for this is #17228.

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 2 above)
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: yeah. it should be after

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: make sense

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: always apply patch at last

**Comment by hzxuzhonghu** [comment]
/retest

TestHttpsAppProbe is not related flake

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 5]
```diff
@@ -121,10 +115,16 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
...
 		}
+		// apply load balancer setting fot cluster endpoints
+		applyLocalityLBSetting(proxy.Locality, outboundClusters, env.Mesh.LocalityLbSetting)
 		outboundClusters = envoyfilter.ApplyClusterPatches(networking.EnvoyFilter_GATEWAY, proxy, push, outboundClusters)
 		clusters = outboundClusters
 	}
 
+	// Add a blackhole and passthrough cluster for catching traffic to unresolved routes
+	// DO NOT CALL PLUGINS for these two clusters.
+	clusters = append(clusters, buildBlackHoleCluster(env), buildDefaultPassthroughCluster(env, proxy))
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: oh this needs to go to top as patch should be applicable to these as well..
the rule of the thumb is that nothing should be changed/added/removed after patch

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: ok

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 6]
```diff
@@ -121,10 +115,16 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
 		if proxy.Type == model.Router && proxy.GetRouterMode() == model.SniDnatRouter {
 			outboundClusters = append(outboundClusters, configgen.buildOutboundSniDnatClusters(env, proxy, push)...)
 		}
+		// apply load balancer setting fot cluster endpoints
```
Comment (johscheuer) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]:
```suggestion
		// apply load balancer setting for cluster endpoints
```

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 7]
```diff
@@ -93,16 +93,10 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
...
-		// apply load balancer setting fot cluster endpoints
-		applyLocalityLBSetting(proxy.Locality, outboundClusters, env.Mesh.LocalityLbSetting)
-	}
-	// Add a blackhole and passthrough cluster for catching traffic to unresolved routes
-	// DO NOT CALL PLUGINS for these two clusters.
-	outboundClusters = append(outboundClusters, buildBlackHoleCluster(env), buildDefaultPassthroughCluster(env, proxy))
-
 	switch proxy.Type {
 	case model.SidecarProxy:
+		// apply load balancer setting fot cluster endpoints
```
Comment (johscheuer) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]:
```suggestion
		// apply load balancer setting for cluster endpoints
```

**Pull request merged**